File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minTillNow = prices[0];
        int maxProfit = 0;
        for(int i=1; i<prices.size(); i++) {
            // try to sell
            int sellPrice = prices[i] - minTillNow;
            maxProfit = max(sellPrice, maxProfit);
            minTillNow = min(minTillNow, prices[i]);
        }
        return maxProfit;
    }
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let maxBuyPrevDays = prices[0];
    let maxProfit = 0;
    for(let i=1; i<prices.length; i++) {
        const profitSellingToday = prices[i] - maxBuyPrevDays;
        if(profitSellingToday > maxProfit) {
            maxProfit = profitSellingToday;
        }
        if(prices[i] < maxBuyPrevDays) {
            maxBuyPrevDays = prices[i];
        }
    }
    return maxProfit;
};